### ISSUE TITLE
`SIM105` should clean up unused variable after refactor

### ISSUE THREAD

**Issue by tylerlaprade** [astral-sh/ruff#6254]
```python
x = None
try:
    x = func_that_might_throw()
except AttributeError:
    pass
return x
```
is getting simplified to
```
x = None  # Variable "x" is not accessed
with contextlib.suppress(AttributeError):
    return func_that_might_throw()
```
If this is the first time we've seen `x`, can we clean up the null assignment?

**Comment by zanieb** [comment]
Thanks for the report! Seems reasonable to me although it may be non-trivial.

**Comment by charliermarsh** [comment]
I believe this works as expected, assuming you enable the right set of rules: `ruff check foo.py --select SIM --select RET --select F -n --fix`.

We like to treat each fix as independent. Here, the `SIM105` rule gives you:

```python
import contextlib
def f():
    x = None
    with contextlib.suppress(AttributeError):
        x = func_that_might_throw()

    return x
```

Then the `RET` rule simplifies to:

```python
import contextlib
def f():
    x = None
    with contextlib.suppress(AttributeError):
        return func_that_might_throw()
```

Then the `F` rule identifies the unused variable:

```python
import contextlib
def f():
    with contextlib.suppress(AttributeError):
        return func_that_might_throw()
```

(This all happens in one `ruff` invocation, I'm just breaking it down for clarity.)

Perhaps you have `F841` disabled?

**Comment by tylerlaprade** [comment]
Ahh, @charliermarsh is correct.
```toml
unfixable = ["B007", "ERA001", "F841", "RUF100"]
```
We disabled autofix on `F841` because variables are often temporarily unused while rearranging a file. We don't want an errant `⌘ + S` to accidentally delete useful code. In this case, though, there's no danger since it's only unused as a result of the other (safe) autofixes.

**Comment by zanieb** [comment]
@tylerlaprade I agree that seems like a fair use-case. We probably don't want to resolve this _just_ for this specific rule though, instead we should work on a general solution that checks for unused variables _introduced_ by an autofix.

Would you mind if I retitle this issue?

**Comment by tylerlaprade** [comment]
Yes, please do! I'm glad this could potentially turn out to be more generally applicable than just a single rule.

**Comment by zanieb** [comment]
In the meantime, for this use-case, I'd recommend leaving `F841` enabled in your configuration and only disabling it in the flags passed when used by the IDE. Would that help?

**Comment by tylerlaprade** [comment]
Yes, that's a good suggestion, thank you Zanie!

**Comment by tylerlaprade** [comment]
Ah, no luck - that same flag is used for the recursive fixes, so `⌘ + S` still has the same behavior.

**Comment by charliermarsh** [comment]
I think the suggestion was to leave you with a setup such that unused variables would be removed when running via the CLI, but not when running in the editor (which could happen "by accident").
